.ycm_extra_conf.py

```diff
@@ -0,0 +1,182 @@
+#
+# This is free and unencumbered software released into the public domain.
+#
+# Anyone is free to copy, modify, publish, use, compile, sell, or
+# distribute this software, either in source code form or as a compiled
+# binary, for any purpose, commercial or non-commercial, and by any
+# means.
+#
+# In jurisdictions that recognize copyright laws, the author or authors
+# of this software dedicate any and all copyright interest in the
+# software to the public domain. We make this dedication for the benefit
+# of the public at large and to the detriment of our heirs and
+# successors. We intend this dedication to be an overt act of
+# relinquishment in perpetuity of all present and future rights to this
+# software under copyright law.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# For more information, please refer to <http://unlicense.org/>
+
+# This config requires that YCM was built with clangd support, which is the
+# recommended configuration today.  See
+# https://github.com/ycm-core/YouCompleteMe#c-family-semantic-completion
+
+from distutils.sysconfig import get_python_inc
+import glob
+import os
+import platform
+import subprocess
+import ycm_core
+
+DIR_OF_THIS_SCRIPT = os.path.abspath(os.path.dirname(__file__))
+SOURCE_EXTENSIONS = [".cpp", ".cxx", ".cc", ".c", ".m", ".mm"]
+
+
+def pio_pkg_dir(pkg):
+    return os.path.join(os.environ["HOME"], ".platformio/packages", pkg)
+
+
+# We manually specify compile flags, which is a bummer.  It's theoretically
+# possible to use compile_commands.json; platformio can generate it,
+# https://docs.platformio.org/en/latest/integration/compile_commands.html.  But
+# there are a few problems.
+#
+#  - platformio does not -I libraries that are not "currently used".  This
+#    means that when you add a new header to a file, ycm won't know how to use
+#    that header until you manually regenerate compile_commands.json!
+#
+#  - platformio uses gcc-avr to compile for Arduino, and that compiler
+#    implicitly adds a bunch of necessary flags that ycm/clangd are not aware
+#    of.  So we'd have to add them here *anyway*.
+#
+#  - Most files get compiled with Arduino flags, but some (tests) must be
+#    compiled with native flags.  So even if we used a compilation_database,
+#    we'd have to intelligently choose which set of flags to use here.
+#
+# Flags are defined below.  Feel free to modify them if necessary!
+
+common_flags = ["-x", "c++", "-std=gnu++17", "-fno-exceptions", "-Wall",] + [
+    "-I" + d for d in {os.path.dirname(h) for h in glob.glob("**/*.h", recursive=True)}
```

Comment:
Does this glob end up including all of our headers, or all of the headers under .pio too? Please add a comment either way.

Reply:
It doesn't include `.pio/`; top of https://docs.python.org/3.8/library/glob.html (although, hilariously, it doesn't say *how* filenames beginning with `.` are a special case).